### PR TITLE
Multiple `catch`'es

### PR DESCRIPTION
**Pull request by sjinks**
See #1336

### CODE REVIEW

File: Library/Statements/TryCatchStatement.php [hunk 1]
```diff
@@ -97,28 +101,34 @@ public function compile(CompilationContext $compilationContext)
...
-
-                    $ifStatement = new IfStatement($ifCheck->build());
-                    $ifStatement->compile($compilationContext);
                 }
+            }
 
-                if ($variable->isTemporal()) {
-                    $variable->setIdle(true);
-                }
+            $primaryIf = $ifs[0];
```
Comment (sjinks) [inline Library/Statements/TryCatchStatement.php:132]: Is there a cleaner way to do this?

Right now this generates something like:

```с
if (...) {
} else {
    if (...) {
    } else {
        if (...) {
            ...
        }
    }
}

File: Library/Statements/TryCatchStatement.php [hunk 2]
```diff
@@ -97,28 +101,34 @@ public function compile(CompilationContext $compilationContext)
...
-                                $exprBuilder->variable($variable->getName()),
+                                $exprBuilder->variable($exc_var->getName()),
                                 $exprBuilder->variable($class['value'])
                             )
                         )
                         ->setStatements($exprBuilder->statements()->block(array_merge(
-                            array($exprBuilder->statements()->rawC('zend_clear_exception(TSRMLS_C);')),
+                            array(
+                                $exprBuilder->statements()->rawC('zend_clear_exception(TSRMLS_C);'),
+                                $exprBuilder->statements()->rawC('ZEPHIR_CPY_WRT(&' . $variable->getName() . ', &' . $exc_var->getName() . ');'),
```
Comment (sjinks) [inline Library/Statements/TryCatchStatement.php]: Is there a cleaner way?

`$compilationContext->backend->copyOnWrite()` writes the code, what I need is the assignment operator as this probably won't compile on ZE2

File: unit-tests/Extension/ExceptionsTest.php [hunk 3]
```diff
@@ -143,4 +144,37 @@ public function testExceptionRethrow()
             $this->assertSame(8, $e->getLine());
```
Comment (sjinks) [inline unit-tests/Extension/ExceptionsTest.php]: Just wonder why line number is 8.

The test says it is 11, I thought it should be 83.

@Jurigag any idea?

File: Library/Statements/TryCatchStatement.php (hunk 1 above)
Comment (Jurigag) [reply Library/Statements/TryCatchStatement.php:132]: Well, the best would be somehow to introduce elseif somehow obviously.

File: unit-tests/Extension/ExceptionsTest.php (hunk 3 above)
Comment (Jurigag) [reply unit-tests/Extension/ExceptionsTest.php]: https://github.com/phalcon/zephir/blob/master/test/exceptions.zep#L8 this is this eight line though as i see there in code maybe you are correct, it should be most likely https://github.com/phalcon/zephir/blob/master/test/exceptions.zep#L80 ?

**Comment by sergeyklay** [comment]
Looks good to me

File: unit-tests/Extension/ExceptionsTest.php (hunk 3 above)
Comment (sjinks) [reply unit-tests/Extension/ExceptionsTest.php]: Make sense… Your test patch added 3 lines to the top of test/exceptions.zep this is why the number is 11 now

**Comment by sergeyklay** [comment]
As I can see there is code generation issue

File: Library/Statements/TryCatchStatement.php [hunk 4]
```diff
@@ -85,7 +89,11 @@ public function compile(CompilationContext $compilationContext)
                     $variable = $compilationContext->symbolTable->getTempVariableForWrite('variable', $compilationContext, $compilationContext);
                 }
 
-                $compilationContext->backend->copyOnWrite($variable, 'EG(exception)', $compilationContext);
+                if ($compilationContext->backend->isZE3()) {
+                    $assignExceptionVarStmt = $exprBuilder->statements()->rawC('ZEPHIR_CPY_WRT(&' . $variable->getName() . ', &' . $exc_var->getName() . ');');
+                } else {
+                    $assignExceptionVarStmt = $exprBuilder->statements()->rawC('ZEPHIR_CPY_WRT(' . $variable->getName() . ', ' . $exc_var->getName() . ');');
+                }
```
Comment (sjinks) [inline Library/Statements/TryCatchStatement.php:97]: The above is a dirty hack, but I failed to make this:

```php
$exprBuilder->statements()->let(array(
     $exprBuilder->operators()->assignVariable($variable->getName(), $exprBuilder->variable($exc_var->getName()))
))->build(),
```

or similar constructs work :-(

If someone knows how to make them work, I would prefer it to this code.

**Comment by sjinks** [comment]
The code works, I am therefore merging it into master, but if someone has a better idea how to handle [this](https://github.com/phalcon/zephir/pull/1450#pullrequestreview-30360486), that would be great.

File: Library/Statements/TryCatchStatement.php (hunk 4 above)
Comment (sergeyklay) [reply Library/Statements/TryCatchStatement.php:97]: Cc: @ovr

**Pull request merged**